### PR TITLE
when window gains focus, update mouse coordinate

### LINKED ISSUE #11130: When the Servo window gains focus, it should update the mouse coordinate
This is important for browserhtml. We need this to build automated tests.

in `ports/glutin`, the mouse coordinate is updated when receiving a glutin mousemove event.

When Servo starts, as the mouse moves, `mouse_pos` is updated. Then Servo loses focus. User moves the mouse. When Servo gains focus, the mouse is likely not at `mouse_pos` anymore. That means:
1. `mouseover` is not triggered and `:hover` styles are not applied for element below the mouse
2. if a click occurs before the mouse is moved, the clicked is triggered for the element that was under the latest known coordinate (right before losing focus)

We should update `mouse_pos` when gaining focus, and fire a mousemove event if it changed.

There is a `Focused(bool)` glutin event, but it doesn't come with new mouse coordinates.

### PR DESCRIPTION
**Pull request by mrmiywj**
---

- [ ] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11130 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because these cannot be tested automated.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11794)

### CODE REVIEW

**Comment by mrmiywj** [comment]
cc @paulrouget 
Since [my modify to servo/glutin](https://github.com/servo/glutin/pull/97) has not been merged in crate.io. This PR cannot compiled successfully.

**Comment by KiChjang** [comment]
You can update glutin by modifying the Cargo.toml file under ports/glutin and running `./mach update-cargo -p glutin`

File: ports/glutin/window.rs [hunk 1]
```diff
@@ -261,11 +261,21 @@ impl Window {
...
-                                    mouse_button == MouseButton::Right {
-                        let mouse_pos = self.mouse_pos.get();
-                        self.handle_mouse(mouse_button, element_state, mouse_pos.x, mouse_pos.y);
+                    mouse_button == MouseButton::Right {
+                        match pos {
+                            Some(x, y) => {
+                                self.mouse_pos.set(Point2D::new(x, y));
+                                self.handle_mouse(mouse_button, element_state, x, y);
+                                self.event_queue.borrow_mut().push(
+                                WindowEvent::MouseWindowMoveEventClass(Point2D::typed(x as f32, y as f32)));
```
Comment (paulrouget) [inline ports/glutin/window.rs]: You need a move event, but also a click event. So you'll need to call `handle_mouse` here as well.

File: ports/glutin/Cargo.toml [hunk 2]
```diff
@@ -17,7 +17,7 @@ log = "0.3.5"
 msg = {path = "../../components/msg"}
 net_traits = {path = "../../components/net_traits"}
 script_traits = {path = "../../components/script_traits"}
-servo-glutin = "0.4"
+servo-glutin = { git = "https://github.com/servo/glutin"}
```
Comment (Manishearth) [inline ports/glutin/Cargo.toml]: https://github.com/servo/glutin/pull/98

File: ports/glutin/window.rs (hunk 1 above)
Comment (mrmiywj) [inline ports/glutin/window.rs]: Hmm, I'm wrong here. But I do think the solution should be:
1. set the new position
2. push the `MouseWindowMoveEventClass` into `event_queue`
3. call `handle_mouse`

For the move event, should I also call `handle_mouse`? I though `handle_mouse` is only for the click event.

Comment (paulrouget) [inline ports/glutin/window.rs]: I believe the mousemove event should happen when the window gains focus.
See my latest comment.

**Comment by paulrouget** [comment]
STR:
1. load the above example
2. click anywhere. see the red div being moved exactly where the click occurs
3. move the mouse anywhere above the window. Do not click. Let's call this new position P1
4. with Alt-Tab (or Cmd-Tab), bring another non-Servo window above the Servo window (make it so the servo window loses focus)
5. move the mouse 10 pixels to the right. Let's call this new position P2
6. with Alt-Tab (or Cmd-Tab), bring back the Servo window. Do not move mouse.
7. Click
8. the red div moves to P1, not to P2. Which is wrong.

Ideally, we should get a mouse move event when 6) happens.

```html
<style>
  div {
    width: 0px;
    height: 0px;
    position: absolute;
    border: 2px solid red;
    top: 0;
    left: 0;
  }
</style>
<div></div>
<script>
  var div = document.querySelector("div");
  window.addEventListener("click", e => {
    div.style.top = e.clientY - 2 + "px";
    div.style.left = e.clientX - 2 + "px";
  });
</script>
```

**Comment by mrmiywj** [comment]
@paulrouget 
Hi, I've updated my PR. I do think it's enough.
But unfortunately, it cannot be compiled on my RMBP.

```
Updating git repository `https://github.com/servo/glutin`
error: no matching package named `servo-glutin` found (required by `glutin_app`)
location searched: https://github.com/servo/glutin
version required: *
Build completed in 0:00:07
```
Is there anything wrong with my modification to `Cargo.toml`?

**Comment by jdm** [comment]
The change to gluten will need to be published, and then we'll need to use the version from crates.io instead of git.

**Comment by mrmiywj** [comment]
@jdm 
So what do I need to do now? My PR is merged by [servo/glutin](https://github.com/servo/glutin/pull/97), which is forked.

**Comment by Manishearth** [comment]
`./mach update-cargo -p servo-glutin` , and commit the lockfile changes

**Comment by mrmiywj** [comment]
@Manishearth 
```
➜  servo git:(update-mouse-coordinate-when-focus) ✗ ./mach update-cargo -p servo-glutin
components/servo
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Updating git repository `https://github.com/servo/glutin`
error: no matching package named `servo-glutin` found (required by `glutin_app`)
location searched: https://github.com/servo/glutin
version required: *
ports/cef
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Updating git repository `https://github.com/servo/glutin`
error: no matching package named `servo-glutin` found (required by `glutin_app`)
location searched: https://github.com/servo/glutin
version required: *
ports/geckolib
error: package id specification `servo-glutin` matched no packages
```
It also complains. QAQ

**Comment by Manishearth** [comment]
Oh, undo your check to that cargo.toml too. Mention glutin 0.4 or 0.4.24

**Comment by mrmiywj** [comment]
Sorry for that maybe I cannot understand what you said.
Maybe u said is that I should specify the `servo-glutin` version based on my commit? Like this:
`servo-glutin = { git = "https://github.com/servo/glutin", version = "0.4.24"}`
But it also complained...
```
➜  servo git:(update-mouse-coordinate-when-focus) ✗ ./mach update-cargo -p servo-glutin
components/servo
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Updating git repository `https://github.com/servo/glutin`
error: no matching package named `servo-glutin` found (required by `glutin_app`)
location searched: https://github.com/servo/glutin
version required: ^0.4.24
ports/cef
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Updating git repository `https://github.com/servo/glutin`
error: no matching package named `servo-glutin` found (required by `glutin_app`)
location searched: https://github.com/servo/glutin
version required: ^0.4.24
ports/geckolib
error: package id specification `servo-glutin` matched no packages
```

Or just like the `Cargo.toml` in master? Like:
`servo-glutin="0.4.24"`
But it will use the repo on [crate.io](crate.io), not our own repo, I think.
It also complained...
```
➜  servo git:(update-mouse-coordinate-when-focus) ✗ ./mach update-cargo -p servo-glutin
components/servo
    Updating registry `https://github.com/rust-lang/crates.io-index`
    Updating cocoa v0.4.3 -> v0.4.4
    Updating servo-glutin v0.4.23 -> v0.4.24
ports/cef
    Updating registry `https://github.com/rust-lang/crates.io-index`
error: failed to select a version for `cocoa` (required by `servo-glutin`):
all possible versions conflict with previously selected versions of `cocoa`
  version 0.4.3 in use by cocoa v0.4.3
  possible versions to select: 0.4.4
ports/geckolib
error: package id specification `servo-glutin` matched no packages
```

File: ports/glutin/Cargo.toml (hunk 2 above)
Comment (Manishearth) [inline ports/glutin/Cargo.toml]: Just change this line to "0.4.24"

**Comment by jdm** [comment]
You may need to run `./mach cargo-update -p cocoa` first.

**Comment by mrmiywj** [comment]
I've updated this PR.
But if [this PR](https://github.com/servo/glutin/pull/97) is not merged by the [crate.io](crate.io). I do think it will not compile successfully...

**Comment by mrmiywj** [comment]
I've corrected what related to this PR.
But there seems some other errors...
```
➜  servo git:(update-mouse-coordinate-when-focus) ✗ ./mach build --dev
   Compiling glutin_app v0.0.1 (file:///Users/ivan/Documents/code/servo/ports/glutin)
   Compiling layout v0.0.1 (file:///Users/ivan/Documents/code/servo/components/layout)
/Users/ivan/Documents/code/servo/ports/glutin/window.rs:84:13: 84:43 error: no method named `with_transparent_corner_radius` found for type `glutin::WindowBuilder<'_>` in the current scope
/Users/ivan/Documents/code/servo/ports/glutin/window.rs:84            .with_transparent_corner_radius(8)
                                                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
Build failed, waiting for other jobs to finish...
error: Could not compile `glutin_app`.

To learn more, run the command again with --verbose.
Build completed in 0:01:00
```

**Comment by jdm** [comment]
Looks like there were changes made to glutin but nobody updated Servo to use them.

**Comment by paulrouget** [comment]
> Looks like there were changes made to glutin but nobody updated Servo to use them.

Right. I think we should just remove the call to `.with_transparent_corner_radius(8)`.
See https://github.com/servo/glutin/pull/95

@pcwalton we need to update glutin to pickup some changes. Can you confirm that we only need to remove the call to `.with_transparent_corner_radius(8)`

**Comment by mrmiywj** [comment]
Actually it only needs to remove the call to `.with_transparent_corner_radius(8)`. Now it works well on mac.

**Comment by pcwalton** [comment]
Yes, just remove that call.

**Comment by jdm** [comment]
@mrmiywj If you rebase this I will merge it. Sorry this got lost in the shuffle.

**Comment by mrmiywj** [comment]
@jdm I've rebase it.

File: ports/glutin/window.rs [hunk 3]
```diff
@@ -260,12 +260,21 @@ impl Window {
...
+                                self.mouse_pos.set(Point2D::new(x, y));
+                                self.event_queue.borrow_mut().push(
+                                    WindowEvent::MouseWindowMoveEventClass(Point2D::typed(x as f32, y as f32)));
+                                self.handle_mouse(mouse_button, element_state, x, y);
+                            }
+                            None => {
+                                let mouse_pos = self.mouse_pos.get();
+                                self.handle_mouse(mouse_button, element_state, mouse_pos.x, mouse_pos.y);
+                            }
+                        }
```
Comment (KiChjang) [inline ports/glutin/window.rs]: I believe this is at a wrong level of indentation. The last closing brace doesn't match up with the `if`.

Comment (mrmiywj) [inline ports/glutin/window.rs]: Correct it now.

File: ports/glutin/window.rs [hunk 4]
```diff
@@ -279,7 +288,8 @@ impl Window {
                     MouseScrollDelta::PixelDelta(dx, dy) => (dx, dy),
                 };
                 let phase = glutin_phase_to_touch_event_type(phase);
-                self.scroll_window(dx, dy, phase);
+                self.scroll_window(
+dx, dy, phase);
```
Comment (KiChjang) [inline ports/glutin/window.rs]: Revert this change please.

File: ports/glutin/window.rs [hunk 5]
```diff
@@ -260,13 +260,22 @@ impl Window {
...
             }
-            Event::MouseInput(element_state, mouse_button, _) => {
-                // FIXME(#11130, pcwalton): The third field should be used.
+            Event::MouseInput(element_state, mouse_button, pos) => {
                 if mouse_button == MouseButton::Left ||
-                                    mouse_button == MouseButton::Right {
-                        let mouse_pos = self.mouse_pos.get();
-                        self.handle_mouse(mouse_button, element_state, mouse_pos.x, mouse_pos.y);
-                   }
+                    mouse_button == MouseButton::Right {
```
Comment (KiChjang) [inline ports/glutin/window.rs]: I think the problem is here, the `mouse_button` here should match up with the last line's `mouse_button` columns, and the `match` expression on the next line should be de-indented once.

**Comment by mrmiywj** [comment]
@KiChjang Is the indent correct now?

File: ports/glutin/window.rs [hunk 6]
```diff
@@ -260,13 +260,22 @@ impl Window {
...
+                               self.event_queue.borrow_mut().push(
+                                   WindowEvent::MouseWindowMoveEventClass(Point2D::typed(x as f32, y as f32)));
+                               self.handle_mouse(mouse_button, element_state, x, y);
+                           }
+                           None => {
+                               let mouse_pos = self.mouse_pos.get();
+                               self.handle_mouse(mouse_button, element_state, mouse_pos.x, mouse_pos.y);
+                           }
+                       }
+                }
```
Comment (KiChjang) [inline ports/glutin/window.rs]: The last one is correct, but you can see that the one above it is indented too much to the right. De-indenting once should fix it.

**Comment by mrmiywj** [comment]
I think it's due to the two lines of if conditions, which may mistake my emacs auto-intending.

**Comment by KiChjang** [comment]
@bors-servo r+

Thank you!
